### PR TITLE
Make openedx.yaml process idempotent, so that it can be resumable

### PR DESCRIPTION
**Pull request by cpennington**
@nedbat: Review?

### CODE REVIEW

**Comment by nedbat** [comment]
@cpennington looks like this PR got clobbered.

File: edx_repo_tools/oep2/explode_repos_yaml.py [hunk 1]
```diff
@@ -48,39 +49,93 @@ def explode(hub, dry):
...
+        else:
+            if not dry:
+                pull = gh_repo.create_pull(
+                    title=pr_title,
+                    base=gh_repo.default_branch,
+                    head=BRANCH_NAME,
+                    body=pr_body
+                )
+            click.secho('{}reated pull request {} against {}'.format(
+                'Would have c' if dry else 'C',
```
Comment (nedbat) [inline edx_repo_tools/oep2/explode_repos_yaml.py]: whoa.

File: edx_repo_tools/oep2/explode_repos_yaml.py [hunk 2]
```diff
@@ -48,39 +49,93 @@ def explode(hub, dry):
...
-            pull.html_url,
-            repo,
-        ), fg='green')
+        if gh_repo.fork:
+            LOGGER.info("Skipping %s because it is a fork", gh_repo.full_name)
+            continue
+
+        try:
+            parent_commit = gh_repo.default_branch.commit.sha
+        except:
```
Comment (nedbat) [inline edx_repo_tools/oep2/explode_repos_yaml.py]: `except Exception:`

File: edx_repo_tools/oep2/explode_repos_yaml.py [hunk 3]
```diff
@@ -48,39 +49,93 @@ def explode(hub, dry):
...
+            LOGGER.warning(
+                "No commit on default branch %s in repo %s",
```
Comment (nedbat) [inline edx_repo_tools/oep2/explode_repos_yaml.py]: When does this happen? If it does happen, `parent_commit` will be undefined.

File: edx_repo_tools/oep2/explode_repos_yaml.py [hunk 4]
```diff
@@ -48,39 +49,93 @@ def explode(hub, dry):
...
+            This adds an `openedx.yaml` file, as described by OEP-2:
+            http://open-edx-proposals.readthedocs.io/en/latest/oeps/oep-0002.html
+
+            The data in this file was transformed from the contents of
+            edx/repo-tools-data:repos.yaml
+        """)
+        pr_title = 'Add an OEP-2 compliant openedx.yaml file',
+
+        existing_pr = [
+            pr
```
Comment (nedbat) [inline edx_repo_tools/oep2/explode_repos_yaml.py]: Wouldn't this be easier?  `list(gh_repo.iter_pulls(...))`.

Or: `existing_pr = next(gh_repo.iter_pulls(...), None)`

**Comment by nedbat** [comment]
A few comments.

File: edx_repo_tools/oep2/explode_repos_yaml.py (hunk 3 above)
Comment (cpennington) [inline edx_repo_tools/oep2/explode_repos_yaml.py]: Continue moves on to the next repo.

Comment (cpennington) [inline edx_repo_tools/oep2/explode_repos_yaml.py]: Oh, sorry. That's a change I have locally (adding a `continue` to the except block).

**Comment by nedbat** [comment]
Is this work stable now, that is, are you waiting for thumbs?

**Comment by cpennington** [comment]
@nedbat: Yes.

File: edx_repo_tools/oep2/explode_repos_yaml.py [hunk 5]
```diff
@@ -5,16 +5,25 @@
...
 from edx_repo_tools.auth import pass_github
 
+logging.basicConfig()
 LOGGER = logging.getLogger(__name__)
 
 BRANCH_NAME = 'add-openedx-yaml'
 OPEN_EDX_YAML = 'openedx.yaml'
 
 
+def dry_echo(dry, message, *args, **kwargs):
```
Comment (nedbat) [inline edx_repo_tools/oep2/explode_repos_yaml.py]: You can never have too many docstrings.

Comment (cpennington) [inline edx_repo_tools/oep2/explode_repos_yaml.py]: True, but I often have too few. =)

File: edx_repo_tools/oep2/explode_repos_yaml.py [hunk 6]
```diff
@@ -48,62 +57,181 @@ def explode(hub, dry):
...
     """
     Implode all openedx.yaml files, and print the results as formatted output.
     """
-    data = dict(iter_openedx_yaml(hub, org))
+    data = dict(iter_openedx_yaml(hub, org, branch))
     click.echo(yaml.safe_dump(data, encoding=None, indent=4))
 
 
-def iter_openedx_yaml(hub, orgs):
+def iter_openedx_yaml(hub, orgs, branches=None):
```
Comment (nedbat) [inline edx_repo_tools/oep2/explode_repos_yaml.py]: Sounds like an interesting function, I wonder what it does? :)

File: edx_repo_tools/oep2/explode_repos_yaml.py [hunk 7]
```diff
@@ -48,62 +57,181 @@ def explode(hub, dry):
...
+                    title=pr_title,
+                    base=gh_repo.default_branch,
+                    head=BRANCH_NAME,
+                    body=pr_body
+                )
 
 
 @click.command()
 @pass_github
 @click.option('--org', multiple=True, default=['edx', 'edx-ops'])
```
Comment (nedbat) [inline edx_repo_tools/oep2/explode_repos_yaml.py]: I wonder if the default should include edx-solutions.

Comment (nedbat) [inline edx_repo_tools/oep2/explode_repos_yaml.py]: And BTW, I can only see one repo in edx-ops, so I wonder if it should be in the default also.

File: edx_repo_tools/oep2/explode_repos_yaml.py [hunk 8]
```diff
@@ -48,62 +57,181 @@ def explode(hub, dry):
...
-            if contents is None:
-                LOGGER.debug("Skipping %s because there is no %s", repo.full_name, OPEN_EDX_YAML)
-                continue
+            for branch in branches:
+                contents = repo.contents(OPEN_EDX_YAML, ref=branch)
+                if contents is not None:
+                    LOGGER.debug("Found openedx.yaml at %s:%s", repo.full_name, branch)
+                    yield repo.full_name, yaml.safe_load(contents.decoded)
+                    break
+            else:
```
Comment (nedbat) [inline edx_repo_tools/oep2/explode_repos_yaml.py]: Seems like this else would be better handled by appending repo.default_branch to branches, and then letting the loop handle it.

File: edx_repo_tools/oep2/explode_repos_yaml.py [hunk 9]
```diff
@@ -48,62 +57,181 @@ def explode(hub, dry):
...
 
 @click.command()
 @pass_github
 @click.option('--org', multiple=True, default=['edx', 'edx-ops'])
-def implode(hub, org):
+@click.option(
+    '--branch',
+    multiple=True,
+    default=None,
+    help="Which branches to look at for openedx.yaml contents. Earlier branches will get priority"
```
Comment (nedbat) [inline edx_repo_tools/oep2/explode_repos_yaml.py]: I thought "Earlier" meant, "created earlier", which in retrospect was silly.  Maybe, "The branch(es) to examine for openedx.yaml files.  If more than one, first found is used."

**Comment by nedbat** [comment]
Reviewed.

**Comment by cpennington** [comment]
@shaunagm: This is the first of my PRs.

File: edx_repo_tools/oep2/explode_repos_yaml.py [hunk 10]
```diff
@@ -213,25 +229,33 @@ def implode(hub, org, branch):
...
 def iter_openedx_yaml(hub, orgs, branches=None):
+    """
+    Yield the data from all openedx.yaml files found in repositories in ``orgs``
+    on any of ``branches``.
+
+    Arguments:
+        hub (GitHub): A connection to GitHub.
+        orgs: A list of github orgs to search for openedx.yaml files.
+        branches: A list of branches to search for openedx.yaml files. If
+            a that file exists on multiple branches, then only the contents
```
Comment (nedbat) [inline edx_repo_tools/oep2/explode_repos_yaml.py]: There's an extra "a".

File: edx_repo_tools/oep2/explode_repos_yaml.py [hunk 11]
```diff
@@ -213,25 +229,33 @@ def implode(hub, org, branch):
...
+            of the first will be yielded. The repositories default branch will
```
Comment (nedbat) [inline edx_repo_tools/oep2/explode_repos_yaml.py]: ```The repository's default branch```

**Comment by nedbat** [comment]
Only two new micro-nits. :+1:

**Pull request merged**